Dash Template (#33)
dwave-examples/template: README.md

**Interested in contributing a code example?** 

Please take a look at our [contribution guidelines](CONTRIBUTING.md) before

> getting started. Thank you!

Reviewer: Does the CONTRIBUTING need to be updated for Dash examples? 
```
Tests that are discoverable through python -m unittest discover
```
Dash uses pytest

following questions:

* Is it pedagogical or a usable application?
* Does it belong to a particular domain such as material simulation or logistics?

> * What level of Ocean proficiency does it target (beginner, advanced, professional)?

Reviewer: What's actually the distinction between "advanced" and "professional"?

Author: I have no idea & I don't think @thisac does either as we had a conversation about this a week or two ago.

Maybe someone else knows the secret? @randomir or @arcondello? 

If we don't already, it would be good to have a doc outlining the difference between the levels (if we plan on keeping them).

Reviewer 2: We're currently investigating the demo filters in Leap and will most likely be updating this further after we've decided if, and how, we'd like to change the tags. This doesn't seem to have anything directly to do with those, though, so I'd suggest we change this to just include beginner and advanced (not sure either what the distinction would be between advanced and professional), or just leave it for now and rethink this in a future PR.

A clear description allows us to properly categorize your example.

Images are encouraged. If your example produces a visualization, consider
displaying it here.

![D-Wave Logo](static/dwave_logo.svg)

> ## Installation

Reviewer: We're switching between instructions to the example contributor and boilerplate text for example users without distinguishing the audiences. Should we use comments or something before sections that the example contributor should keep as is?

You can run this example without installation in cloud-based IDEs that support the [Development Containers specification](https://containers.dev/supporting) (aka "devcontainers").

For development environments that do not support `devcontainers`, install requirements:

```bash
pip install -r requirements.txt
```

If you are cloning the repo to your local system, working in a [virtual environment](https://docs.python.org/3/library/venv.html) is recommended.

## Usage
Your development environment should be configured to [access Leap’s Solvers](https://docs.ocean.dwavesys.com/en/stable/overview/sapi.html). You can see information about supported IDEs and authorizing access to your Leap account [here](https://docs.dwavesys.com/docs/latest/doc_leap_dev_env.html).

### Inputs
If your example requires user input, make sure to specify the input format and any input limitations.
### Outputs
An example program output.

> ## Demo Types

Reviewer: I don't understand the purpose of this Header 1: For a Demo Types section, I'd expect more than one type of demo to be listed/explained. The next line presents the Dash type as a Header 2 but the next Header 2 is Problem Description
Is the intention to add another subsection for non-Dash templates?

Author: Yes, I added this a bit prematurely as the plan is to add a Jupyter Notebook version and a script version. If you think it would be clearer I can remove it for the time being.

> ### User Interface Demo (Dash Framework Template)

Reviewer: Might be clearer to simply have "### Demos Using Dash Interface"

Presumably a user contributing to our examples will know we use Dash for fancy GUIs so this will be easily understood

> The Dash template is intended for demos that would benefit from a user interface. This user interface could include settings to run and customize the problem, an interactive graphical element, and/or tables or charts to compare different solutions. This template is also useful for demos that are intended for a general audience, as it is more approachable for those without a technical background.

Reviewer: For text like this we really need a way to distinguish two types of content in this template:
1. Explanations of the template directed at the contributor 
2. Content that should be in the new example's README

The way we do this in PS templates for customer reports is with color but GitHub support for colored text is [complicated](https://github.com/orgs/community/discussions/31570). Perhaps a standard sentence placed before content that should be in the example README

Author: Yeah I was struggling with this too while writing it. Do you think it would be helpful to separate the two sections into two top level headers that describe that one is a template to fill out and the other is a demo creator manual can be deleted after the demo is done? That way the creator can easily remove one of the two sections and everything is in one place?

Reviewer: That sounds like what might be a good approach. It's a bit hard to know without looking at the result but I think it might work well.

Reviewer 2: I was thinking about this as well, and I think separating them out into two very distinct sections works as a solution, another one being having a separate setup-instructions file. Not sure which I like best.

Author: Ok sounds good I'll give it a shot and you can let me know if you think it's good!

Run the following terminal command to start the Dash app:

```bash
python app.py
```

Access the user interface with your browser at http://127.0.0.1:8050/.

#### Files
* [app.py](app.py) contains all the [Dash callback functions](https://dash.plotly.com/basic-callbacks) required to run the Dash app. Any new callbacks should be added to this file.
* [dash_html.py](dash_html.py) contains all the Dash HTML components that generate the UI for the app, including settings like sliders, checkboxes, and text inputs, as well as buttons and tables. If a Dash HTML component needs to be dynamically added to the UI by a Dash callback in `app.py`, it is encouraged to add a function generating this component to `dash_html.py` and call this function from `app.py`.

> * [app_configs.py](app_configs.py) contains all configerations and settings for the demo and is intended as a quick way to customize the demo for a specific audience or use case. This file is intended for users of the demo and should not be cluttered with settings that are not helpful to the user.

Reviewer: ```suggestion
* [app_configs.py](app_configs.py) contains all configurations and settings for the demo and is intended as a quick way to customize the demo for a specific audience or use case. This file is intended for users of the demo and should not be cluttered with settings that are not helpful to the user.
```

typo

> * [z_app.css](z_app.css) contains all custom CSS styling for the demo; any new CSS rules should be added here. Dash reads all files in the `/assets/` directory from top to bottom so `z_app.css` will be the last file to be run and will, therefore, be able to overwrite other styling rules.

Reviewer: ```suggestion
* [z_app.css](z_app.css) contains all custom CSS styling for the demo; any new CSS rules should be added here. Dash reads all files in the `/assets/` directory in alphabetical order so `z_app.css` runs last and overwrites previous styling rules.
```

#### Directories

> * `/assets/` is a Dash specific directory and that must not contain subdirectories and must not be renamed. Dash reads files in this directory from top to bottom, so no imports are necessary. The `base.css` file is a Dash-made file with some basic styling rules; this file should not be edited and its name should not be changed. All CSS files with a `c` prefix are custom CSS files created specifically for this demo template. For maintainability these files should _ideally_ not be altered.

Reviewer: ```suggestion
* `/assets/` is a Dash-specific directory that must not contain subdirectories and must not be renamed. Dash always reads files in this directory in alphabetical order: you do not need to explicitly import content from this directory. The `base.css` file is a Dash-made file with some basic styling rules; this file should not be edited and its name should not be changed. All CSS files with a `c` prefix are customized for this demo template. _Ideally_, these files should  not be altered.
```

> * `/src/` should contain all functional code for the dash demo including solver implementations, Class definitions, etc. This directory is not specific to Dash and therefore subdirectories can be used.

Reviewer: ```suggestion
* `/src/` should contain all functional code for the dash demo including solver implementations, class definitions, etc. This directory is not specific to Dash and therefore subdirectories can be used.
```
